### PR TITLE
feat: add typescript-eslint rule array-type

### PR DESCRIPTION
**Pull request by luhc228**
Ref: https://github.com/oxc-project/oxc/issues/2180

Most test cases which from [eslint-plugin](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/tests/rules/array-type.test.ts) pass, but some don't. At this time I could not figure out. I want to add more test cases after some time.

### CODE REVIEW

**Review by camc314 — COMMENTED** [review]
missing a snapshot file?

File: crates/oxc_linter/src/rules/typescript/array_type.rs [hunk 1]
```diff
@@ -0,0 +1,879 @@
...
+        }
+            ", Some(serde_json::json!([{"default":"generic"}]))),
+        // ("type Unwrap<T> = T extends Array<infer E> ? E : T;", "type Unwrap<T> = T extends (infer E)[] ? E : T;", Some(serde_json::json!([{"default":"array"}]))),
+        // ("type Unwrap<T> = T extends (infer E)[] ? E : T;", "type Unwrap<T> = T extends Array<infer E> ? E : T;", Some(serde_json::json!([{"default":"generic"}]))),
+        // ("type Foo = ReadonlyArray<object>[];", "type Foo = (readonly object[])[];", Some(serde_json::json!([{"default":"array"}]))),
+        ("const foo: Array<new (...args: any[]) => void> = [];", "const foo: (new (...args: any[]) => void)[] = [];", Some(serde_json::json!([{"default":"array"}]))),
+        ("const foo: ReadonlyArray<new (...args: any[]) => void> = [];", "const foo: readonly (new (...args: any[]) => void)[] = [];", Some(serde_json::json!([{"default":"array"}]))),
+    ];
+
+    Tester::new(ArrayType::NAME, pass, fail).expect_fix(fix).test();
```
Comment (Boshen) [inline crates/oxc_linter/src/rules/typescript/array_type.rs:878]: need to call `test_and_snapshot`

**Review by Boshen — CHANGES_REQUESTED** [review]

**Review by Boshen — APPROVED** [review]
Thank you!

**Comment by luhc228** [comment]
clippy error must fixed before merging pr? @Boshen

**Pull request merged**